### PR TITLE
Reduce nr of dynamic allocations

### PR DESCRIPTION
**Pull request by ShFil119**
Do you prefer using const ref or raw ptr?

### CODE REVIEW

File: rwengine/src/ai/CharacterController.cpp [hunk 1]
```diff
@@ -677,8 +677,8 @@ bool Activities::UseItem::update(CharacterObject *character,
     const auto &weapon = world->data->weaponData.at(itemslot);
```
Comment (madebr) [inline rwengine/src/ai/CharacterController.cpp]: What if there is no weapon at itemslot?

Comment (ShFil119) [reply rwengine/src/ai/CharacterController.cpp]: According to standard, `at` should throw exception. It's connected with question first post. Ref cannot be empty - so if we use this option, then modified and invalid assets could crash game. (ofc pointers also require checks)

Question if is it a problem?

File: rwengine/src/objects/ProjectileObject.hpp [hunk 2]
```diff
@@ -26,7 +26,7 @@ class ProjectileObject final : public GameObject {
 
     struct ProjectileInfo {
         ProjectileInfo(ProjectileType p_type, glm::vec3 p_direction,
```
Comment (madebr) [inline rwengine/src/objects/ProjectileObject.hpp]: Should we keep the `p_` prefixes? It is expected that the constructor takes projectile type, direction, velocity, time and weapon data.

**Comment by madebr** [comment]
If I understand correctly, you are sacrificing run time performance (`O(log(N))` -> `O(N)`) to dynamic memory usage (`N x string length`).
Is it worth it? Do you know how large `N` is?

File: rwengine/src/ai/CharacterController.cpp (hunk 1 above)
Comment (madebr) [reply rwengine/src/ai/CharacterController.cpp]: I guess not now.
Later on, malicious scm or a scm fuzzer can crash the game.

**Comment by ShFil119** [comment]
I can return to `std::map`. (or use unordered_map)

If I start debug mode, then there are 168 dynamic allocations. Map will cause more jumping on memory and duplicating memory usage for key(it's already in second slot).

It's hard to frankly say what's is better `O(N)` with small constant(thx to contiguous memory) or `O(log(N))` with bigger constant.

**Comment by madebr** [comment]
The `std::string` in the `std::vector` will not necessarily be continuous. `std::string`  will allocate sufficient memory on the heap for longer strings. (smaller strings will benefit from the small string optimization)
If you want to keep the string in the `std::vector` (with sufficient capacity) and no other heap allocations, you should use `char string[MAXSIZE]` or `std::array<char, MAXSIZE>`. (idk how to store strings the C++ way)

**Comment by ShFil119** [comment]
Yeah, `std::string` is just ptr to memory. I was thinking overall about all other Plain Old Data. (but still there is one less jump on memory) ;)

Using char for runtime strings (I think) is downgrade.

**Comment by ShFil119** [comment]
Refactored, now uses std::unordered_map. Also avoided duplication of id string.

File: rwengine/src/objects/VehicleInfo.hpp [hunk 3]
```diff
@@ -103,9 +102,15 @@ struct VehicleInfo {
         bool empty() const {
             return front.empty() && back.empty();
         }
-    } seats;
-};
+    };
```
Comment (madebr) [inline rwengine/src/objects/VehicleInfo.hpp]: nit: You can do
```
struct Seats {
   ...
} seats;
```

**Pull request merged**